File: android/src/main/java/uk/co/darkerwaters/flic_button/Flic2Controller.java
package uk.co.darkerwaters.flic_button;

import android.content.Context;
import android.os.Handler;
import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flic.flic2libandroid.Flic2Button;
import io.flic.flic2libandroid.Flic2ButtonListener;
import io.flic.flic2libandroid.Flic2Manager;
import io.flic.flic2libandroid.Flic2ScanCallback;

public class Flic2Controller {
    // keep the buttons so we can call functions on them later from flutter (via UUID)
    private final Map<String, Flic2Button> buttonsDiscovered = new HashMap<>();

    private final ButtonCallback callback;

    public interface ButtonCallback {
        void onPairedButtonFound(Flic2Button button);
        void onButtonFound(Flic2Button button);
        void onButtonConnected();
        void onButtonDiscovered(String buttonAddress);
        void onButtonScanningStarted();
        void onButtonScanningStopped();
        void onButtonClicked(Flic2Button button, boolean wasQueued, boolean lastQueued, long timestamp, boolean isSingleClick, boolean isDoubleClick, boolean isHold);
        void onError(String error);
    }

    public Flic2Controller(Context context, ButtonCallback callback) {
        // one callback to inform per manager
        this.callback = callback;
        // initialise the manager, don't need to remember it as we can just get it later
        Flic2Manager.initAndGetInstance(context, new Handler());
    }

    public boolean startButtonScanning() {
        // cancel any previous scan
        cancelButtonScan();
        // and start a new one
        callback.onButtonScanningStarted();
        Flic2Manager.getInstance().startScan(new Flic2ScanCallback() {
            @Override
            public void onDiscoveredAlreadyPairedButton(Flic2Button button) {
                // Found an already paired button
                storeButtonData(button);
                // and inform the caller of this state
                callback.onPairedButtonFound(button);
            }
            @Override
            public void onDiscovered(String bdAddr) {
                // Found Flic2, now connecting, inform the caller of this state
                callback.onButtonDiscovered(bdAddr);
            }
            @Override
            public void onConnected() {
                // connecting to a flic button
                callback.onButtonConnected();
            }
            @Override
            public void onComplete(int result, int subCode, Flic2Button button) {
                callback.onButtonScanningStopped();
                if (result == Flic2ScanCallback.RESULT_SUCCESS) {
                    // The button object can now be used, store this
                    storeButtonData(button);
                    // and inform the caller of this state
                    callback.onButtonFound(button);
                } else {
                    callback.onError(String.format("Internal FLic2 Scan Error with result %d, subCode: %d", result, subCode));
                }
            }
        });
        return true;
    }

    public boolean cancelButtonScan() {
        // cancel any scanning in progress
        callback.onButtonScanningStopped();
        try {
            Flic2Manager manager = Flic2Manager.getInstance();
            if (null != manager) {
                manager.stopScan();
                return true;
            }
        }
        catch (Exception e) {
            callback.onError("Failed to stop scan while releasing flick " + e.getMessage());
        }
        return false;
    }

    private void storeButtonData(Flic2Button button) {
        // store this data for later
        synchronized (buttonsDiscovered) {
            buttonsDiscovered.put(button.getUuid(), button);
        }
    }
    
    public List<Flic2Button> getButtonsDiscovered() {
        List<Flic2Button> buttons = Flic2Manager.getInstance().getButtons();
        for (Flic2Button button : buttons) {
            // while we are here, remember the data in case they want to listen to it
            storeButtonData(button);
        }
        return buttons;
    }
    
    public Flic2Button getButtonForAddress(String buttonAddress) {
        return Flic2Manager.getInstance().getButtonByBdAddr(buttonAddress);
    }

    public boolean connectButton(String buttonUuid) {
        // get the button to listen to from our map and then listen to it
        Flic2Button button;
        synchronized (buttonsDiscovered) {
            button = buttonsDiscovered.get(buttonUuid);
        }
        if (null == button) {
            callback.onError("Cannot connect a button as don't recognise the UUID " + buttonUuid);
            return false;
        } else {
            // and connect to the button
            button.connect();
            return true;
        }
    }

    public boolean disconnectButton(String buttonUuid) {
        // get the button to listen to from our map and then listen to it
        Flic2Button button;
        synchronized (buttonsDiscovered) {
            button = buttonsDiscovered.get(buttonUuid);
        }
        if (null == button) {
            callback.onError("Cannot disconnect a button as don't recognise the UUID " + buttonUuid);
            return false;
        } else {
            // and disconnect from the button
            button.disconnectOrAbortPendingConnection();
            return true;
        }
    }

    public boolean forgetButton(String buttonUuid) {
        // get the button to forget
        Flic2Button button;
        synchronized (buttonsDiscovered) {
            button = buttonsDiscovered.get(buttonUuid);
        }
        if (null == button) {
            callback.onError("Cannot forget a button as don't recognise the UUID " + buttonUuid);
            return false;
        } else {
            // and forget this button
            Flic2Manager.getInstance().forgetButton(button);
            return true;
        }
    }
    
    public boolean listenToButton(String buttonUuid) {
        // get the button to listen to from our map and then listen to it
        Flic2Button button;
        synchronized (buttonsDiscovered) {
            button = buttonsDiscovered.get(buttonUuid);
        }
        if (null == button) {
            callback.onError("Cannot to listen to a button as don't recognise the UUID " + buttonUuid);
            return false;
        } else {
            if (button.getConnectionState() == Flic2Button.CONNECTION_STATE_DISCONNECTED) {
                // to listen to a button we need it connected first, let's assume the caller wants this done
                button.connect();
                // there's a function to inform the listeners of this while we are doing this ourselves
                callback.onButtonConnected();
            }
            // we might already be listening to this, try to remove it first
            button.removeListener(buttonListener);
            // and add it back in to listen to each button only once.
            button.addListener(buttonListener);
            return true;
        }
    }

    public boolean stopListeningToButton(String buttonUuid) {
        // get the button to stop listening to from our map and then listen to it
        Flic2Button button;
        synchronized (buttonsDiscovered) {
            button = buttonsDiscovered.get(buttonUuid);
        }
        if (null == button) {
            callback.onError("Cannot stop listening to a button as don't recognise the UUID " + buttonUuid);
            return false;
        } else {
            button.removeListener(buttonListener);
            return true;
        }
    }

    private final Flic2ButtonListener buttonListener = new Flic2ButtonListener() {
        @Override
        public void onButtonSingleOrDoubleClickOrHold(Flic2Button button, boolean wasQueued, boolean lastQueued, long timestamp, boolean isSingleClick, boolean isDoubleClick, boolean isHold) {
            // let the base deal
            super.onButtonSingleOrDoubleClickOrHold(button, wasQueued, lastQueued, timestamp, isSingleClick, isDoubleClick, isHold);
            // and pass this button press from Flic2 on to our application
            callback.onButtonClicked(button, wasQueued, lastQueued, timestamp, isSingleClick, isDoubleClick, isHold);
        }
    };

    public boolean releaseFlic() {
        // cancel any scanning
        cancelButtonScan();
        // release all the flic 2 listeners on the managers
        try {
            Flic2Manager manager = Flic2Manager.getInstance();
            if (null != manager) {
                // we are probably listening to buttons, stop this
                for (Flic2Button button : manager.getButtons()) {
                    try {
                        button.removeListener(buttonListener);
                    } catch (Exception e) {
                        callback.onError("Failed to remove listener on releasing flic " + e.getMessage());
                    }
                }
                return true;
            }
        }
        catch (Exception e) {
            callback.onError("Failed to destroy the flic two instance as it was not initialised " + e.getMessage());
        }
        return false;
    }
}

File: ios/Classes/Flic2Controller.m
#import <Foundation/Foundation.h>
#import "Flic2Controller.h"
#import "Flic2ControllerListener.h"

@ import flic2lib;

@implementation Flic2Controller
{
    id<Flic2ControllerListener> callback;
    NSMutableDictionary* buttonsDiscovered;
}

- (id)initWithListener:(id<Flic2ControllerListener>)callback {
    if (self = [super init]) {
        self->callback = callback;
        self->buttonsDiscovered = [[NSMutableDictionary alloc] init];
        // and initialize the Flic2 singleton
        [FLICManager configureWithDelegate:self buttonDelegate:self background:YES];
    }
    return self;
    
}

- (void)dispose {
    // shut everything down
    self->callback = nil;
    self->buttonsDiscovered = nil;
    FLICManager* manager = [FLICManager sharedManager];
    // cancel any active scanning
    [manager stopScan];
    // clear us as the delegates
    [manager setDelegate:nil];
    [manager setButtonDelegate:nil];
}

- (void)manager:(nonnull FLICManager *)manager didUpdateState:(FLICManagerState)state {
    switch (state)
    {
        case FLICManagerStatePoweredOn:
            // Flic buttons can now be scanned and connected.
            NSLog(@"Bluetooth is turned on");
            break;
        case FLICManagerStatePoweredOff:
            // Bluetooth is not powered on.
            NSLog(@"Bluetooth is turned off");
            break;
        case FLICManagerStateUnsupported:
            // The framework can not run on this device.
            NSLog(@"FLICManagerStateUnsupported");
        default:
            break;
    }
}

- (void)initializeButton:(FLICButton*)button {
    // put this into our map of data then please, for later access
    [buttonsDiscovered setObject:button forKey:button.uuid];
    // setup the button properly then please
    if (button.triggerMode != FLICButtonTriggerModeClickAndDoubleClickAndHold) {
        // change the mode of the button to tell us everything please
        NSLog(@"changing button to inform about all types of button press");
        button.triggerMode = FLICButtonTriggerModeClickAndDoubleClickAndHold;
    }
}

- (void)managerDidRestoreState:(nonnull FLICManager *)manager {
    // The manager was restored and can now be used.
    for (FLICButton *button in manager.buttons) {
        // and set it up then which will also add it to our map as needed later
        [self initializeButton:button];
        // and inform any listeners of this change in state
        if (nil != self->callback) {
            // pass this to the callback then
            [self->callback onPairedButtonFound:button];
        }
    }
}

- (Boolean)startButtonScanning {
    if (nil != self->callback) {
        // pass this to the callback then
        [self->callback onButtonScanningStarted];
    }
    [[FLICManager sharedManager] scanForButtonsWithStateChangeHandler:^(FLICButtonScannerStatusEvent event) {
        // You can use these events to update your UI.
        switch (event)
        {
            case FLICButtonScannerStatusEventDiscovered:
                // discovery doesn't have the address, so let's just leave it alone here
                break;
            case FLICButtonScannerStatusEventConnected:
                // and inform any listeners of this change in state
                if (nil != self->callback) {
                    // pass this to the callback then
                    [self->callback onButtonConnected];
                }
                break;
            case FLICButtonScannerStatusEventVerified:
                break;
            case FLICButtonScannerStatusEventVerificationFailed:
                NSLog(@"The Flic verification failed.");
                break;
            default:
                break;
        }
    } completion:^(FLICButton *button, NSError *error) {
        if (nil != self->callback) {
            // pass this to the callback then
            [self->callback onButtonScanningStopped];
        }
        if (!error) {
            // Listen to all the click types then please by initializing the button here
            [self initializeButton:button];
            // and inform any listeners of this change in state
            if (nil != self->callback) {
                // pass this to the callback then
                [self->callback onButtonDiscovered:button.bluetoothAddress];
            }
        } else {
            NSLog(@"Scanner completed with error: %@", error);
            // inform any listeners of this error
            if (nil != self->callback) {
                // pass this to the callback then
                [self->callback onError:error.localizedDescription];
            }
        }
    }];
    return true;
}

- (Boolean)stopButtonScanning {
    // stop the scanning
    [[FLICManager sharedManager] stopScan];
    // this will return any active scan with an error so we don't need to send a message from here...
    return true;
}

- (NSArray<FLICButton*>*)getFlic2Buttons {
    // just return our list, but in a way that the caller cannot edit
    return [[self->buttonsDiscovered allValues] copy];
}

- (FLICButton*)getButtonForAddress: (NSString*)address {
    if (nil == address) {
        // not good
        return nil;
    }
    // find the button in the array to return
    for (FLICButton* button in [self->buttonsDiscovered allValues]) {
        if ([button.bluetoothAddress isEqualToString:address]) {
            // this is the one
            return button;
        }
    }
    // not found
    return nil;
}

- (FLICButton*)getButtonForUuid: (NSString*)buttonUuid {
    if (nil == buttonUuid) {
        // not good
        return nil;
    }
    // find the button in the array to return
    for (FLICButton* button in [self->buttonsDiscovered allValues]) {
        if ([button.uuid isEqualToString:buttonUuid]) {
            // this is the one
            return button;
        }
    }
    // not found
    return nil;
}

- (Boolean)listenToButton: (NSString*)buttonUuid {
    FLICButton* button = [self getButtonForUuid:buttonUuid];
    if (nil != button) {
        // there's no listen / don't listen in iOS but we can ask for clicks or not, so let's do that
        button.triggerMode = FLICButtonTriggerModeClickAndDoubleClickAndHold;
        return true;
    } else {
        return false;
    }
}

- (Boolean)stopListeningToButton: (NSString*)buttonUuid {
    // can't really do this, stop listening that is...
    FLICButton* button = [self getButtonForUuid:buttonUuid];
    if (nil != button) {
        // there's no listen / don't listen in iOS but we can ask for clicks or not, so let's do that
        // but it also doesn't let us stop listening to everything )O:
        //button.triggerMode = FLICButtonTriggerModeNone;
        return false;
    } else {
        return false;
    }
}

- (Boolean)connectButton: (NSString*)buttonUuid {
    FLICButton* button = [self getButtonForUuid:buttonUuid];
    if (nil != button) {
        // and connect the button
        [button connect];
        return true;
    } else {
        return false;
    }
}

- (Boolean)disconnectButton: (NSString*)buttonUuid {
    FLICButton* button = [self getButtonForUuid:buttonUuid];
    if (nil != button) {
        // and disconnect the button
        [button disconnect];
        return true;
    } else {
        return false;
    }
}

- (Boolean)forgetButton: (NSString*)buttonUuid {
    FLICButton* button = [self getButtonForUuid:buttonUuid];
    if (nil != button) {
        // and forget the button
        [[FLICManager sharedManager] forgetButton:button completion:^(NSUUID * _Nonnull uuid, NSError * _Nullable error) {
            if (!error) {
                // so we can remove this from our map of buttons
                [self->buttonsDiscovered removeObjectForKey:buttonUuid];
            } else {
                // inform any listeners of this error
                NSLog(@"Forget button failed with error: %@", error);
                if (nil != self->callback) {
                    // pass this to the callback then
                    [self->callback onError:error.localizedDescription];
                }
            }
        }];
        return true;
    } else {
        return false;
    }
}

- (void)button:(nonnull FLICButton *)button didDisconnectWithError:(NSError * _Nullable)error {
    // and inform any listeners of this problem
    if (nil != self->callback && nil != error) {
        // pass this to the callback then
        [self->callback onError:error.localizedDescription];
    }
}

- (void)button:(nonnull FLICButton *)button didFailToConnectWithError:(NSError * _Nullable)error {
    // and inform any listeners of this problem
    if (nil != self->callback && nil != error) {
        // pass this to the callback then
        [self->callback onError:error.localizedDescription];
    }
}

- (void)buttonDidConnect:(nonnull FLICButton *)button {
    // be sure to setup the button properly as it connects
    [self initializeButton:button];
    // and inform any listeners of this lovely action (just like android so can't pass the button)
    if (nil != self->callback) {
        // pass this to the callback then that we are connected
        [self->callback onButtonConnected];
    }
}

- (void)buttonIsReady:(nonnull FLICButton *)button {
    // be sure to setup the button properly as it becomes ready
    [self initializeButton:button];
    // and inform any listeners of this lovely action (acts like the android found function)
    if (nil != self->callback) {
        // pass this to the callback then that we are connected
        [self->callback onButtonFound:button];
    }
}

- (void)button:(FLICButton *)button didReceiveButtonClick:(BOOL)queued age:(NSInteger)age {
    if (nil != self->callback) {
        // pass this to the callback then
        [self->callback onButtonClicked:button wasQueued:queued at:age withClicks:1];
    }
}

- (void)button:(FLICButton *)button didReceiveButtonDoubleClick:(BOOL)queued age:(NSInteger)age {
    if (nil != self->callback) {
        // pass this to the callback then
        [self->callback onButtonClicked:button wasQueued:queued at:age withClicks:2];
    }
}

- (void)button:(FLICButton *)button didReceiveButtonHold:(BOOL)queued age:(NSInteger)age {
    if (nil != self->callback) {
        // pass this to the callback then
        [self->callback onButtonClicked:button wasQueued:queued at:age withClicks:3];
    }
}

@end
